include/wil/cppwinrt_helpers.h

```diff
@@ -0,0 +1,214 @@
+    /// @endcond
+
+    //! Resumes coroutine execution on the thread associated with the dispatcher, or throws
+    //! an exception (from an arbitrary thread) if unable. Supported dispatchers are
+    //! Windows.System.DispatcherQueue, Microsoft.System.DispatcherQueue,
+    //! Microsoft.UI.Dispatching.DispatcherQueue, and Windows.UI.Core.CoreDispatcher,
+    //! but you must include the corresponding <winrt/Namespace.h> header before including
+    //! wil\cppwinrt_helpers.h. It is okay to include wil\cppwinrt_helpers.h multiple times:
+    //! support will be enabled for any winrt/Namespace.h headers that were included since
+    //! the previous inclusion of wil\cppwinrt_headers.h.
+    template<typename Dispatcher>
+    [[nodiscard]] auto resume_foreground(Dispatcher const& dispatcher,
+        typename details::dispatcher_traits<Dispatcher>::Priority priority = details::dispatcher_traits<Dispatcher>::Priority::Normal)
+    {
+        using Traits = details::dispatcher_traits<Dispatcher>;
+        using Priority = typename Traits::Priority;
+        using Handler = typename Traits::Handler;
+
+        struct awaitable
+        {
+            awaitable(Dispatcher const& dispatcher, Priority priority) noexcept :
+                m_dispatcher(dispatcher),
+                m_priority(priority)
+            {
+            }
+            bool await_ready() const noexcept { return false; }
+
+            void await_suspend(details::coroutine_handle<> handle)
+            {
+                m_state.handle = handle;
+                Handler handler{ details::dispatcher_handler(&m_state) };
+                try
+                {
+                    // The return value of Schedule is not reliable. Use the dispatcher_handler destructor
+                    // to detect whether the work item failed to run.
+                    Traits::Scheduler::Schedule(m_dispatcher, m_priority, handler);
+                }
+                catch (...)
+                {
+                    m_state.handle = nullptr; // the exception will resume the coroutine, so the handler shouldn't do it
+                    throw;
+                }
+            }
+
+            void await_resume() const
+            {
+                if (m_state.orphaned)
+                {
+                    throw winrt::hresult_error(HRESULT_FROM_WIN32(ERROR_NO_TASK_QUEUE));
+                }
+            }
+
+        private:
+            Dispatcher const& m_dispatcher;
+            Priority const m_priority;
+            details::dispatched_handler_state m_state;
+        };
+        return awaitable{ dispatcher, priority };
+    }
+}
+#endif // C++20 coroutines supported
+#endif // __WIL_CPPWINRT_HELPERS_DEFINED
+
+/// @cond
+#if defined(WINRT_Windows_UI_Core_H) && !defined(__WIL_CPPWINRT_WINDOWS_UI_CORE_HELPERS)
+#define __WIL_CPPWINRT_WINDOWS_UI_CORE_HELPERS
+namespace wil::details
+{
+    template<>
+    struct dispatcher_traits<winrt::Windows::UI::Core::CoreDispatcher>
+    {

```

Comment:
I found a way to avoid having to protect each use. I just declare a dummy implementation of `wil::details::coroutine_handle<T>` if coroutines aren't supported. Then the rest of the code can just compile as if coroutines were supported. (If not, then you get a dummy implementation.)

Reply:
Why not just error if coroutines aren't supported?